[Bug]: Failed to compile `matplotlib 3.5.0` on aarch64, because of false `TERMUX_BUILD_TUPLE`

### Problem description

Failed to install matplotlib 3.5.0 with `pip install --user matplotlib==3.5.0`, because this command:

```shell
./configure \
    '--with-zlib=no' '--with-bzip2=no' '--with-png=no' '--with-harfbuzz=no' '--enable-static' \
    '--disable-shared' '--host=x86_64-pc-linux-gnu'
```

returned non-zero exit status 2. The argument `--host=x86_64-pc-linux-gnu` is obviously abnormal on my aarch64 device.

This argument comes from [these lines](https://github.com/matplotlib/matplotlib/blob/v3.5.0/setupext.py#L627-L629) which firstly introduced in [this commit](https://github.com/matplotlib/matplotlib/commit/80a29ad3db04ce98d20a4b0df7e56888ce23ffc5). It's determined by the config_var `BUILD_GNU_TYPE`.

```python
# Python 3.10.0 (default, Nov 17 2021, 19:57:18) [Clang 12.0.8 (https://android.googlesource.com/toolchain/llvm-project c935d99d7 on linux
# Type "help", "copyright", "credits" or "license" for more information.

>>> import sysconfig
>>> sysconfig.get_config_var("BUILD_GNU_TYPE")
'x86_64-pc-linux-gnu'

>>> for k, v in sysconfig.get_config_vars().items():
...     if isinstance(v, str) and "x86" in v.lower():
...         print((k, v))
...
('BUILD_GNU_TYPE', 'x86_64-pc-linux-gnu')
('CONFIG_ARGS', "'--disable-dependency-tracking' '--prefix=/data/data/com.termux/files/usr' '--libdir=/data/data/com.termux/files/usr/lib' '--sbindir=/data/data/com.termux/files/usr/bin' '--disable-rpath' '--disable-rpath-hack' '--host=aarch64-linux-android' 'ac_cv_file__dev_ptmx=yes' 'ac_cv_file__dev_ptc=no' 'ac_cv_func_wcsftime=no' 'ac_cv_func_ftime=no' 'ac_cv_func_faccessat=no' '--build=x86_64-pc-linux-gnu' '--with-system-ffi' '--with-system-expat' '--without-ensurepip' 'ac_cv_func_linkat=no' 'ac_cv_posix_semaphores_enabled=no' 'ac_cv_buggy_getaddrinfo=no' '--enable-loadable-sqlite-extensions' 'ac_cv_little_endian_double=yes' 'ac_cv_func_sem_open=no' 'ac_cv_func_sem_timedwait=no' 'ac_cv_func_sem_getvalue=no' 'ac_cv_func_sem_unlink=no' '--disable-nls' '--enable-shared' '--enable-static' '--libexecdir=/data/data/com.termux/files/usr/libexec' 'build_alias=x86_64-pc-linux-gnu' 'host_alias=aarch64-linux-android' 'CC=aarch64-linux-android-clang' 'CFLAGS= -fstack-protector-strong -O3' 'LDFLAGS=-L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now -landroid-support -L/home/builder/.termux-build/_cache/android-r23b-api-24-v1/sysroot/usr/lib' 'CPPFLAGS= -I/data/data/com.termux/files/usr/include -I/home/builder/.termux-build/_cache/android-r23b-api-24-v1/sysroot/usr/include' 'CPP=aarch64-linux-android-cpp'")

>>> for arg in shlex.split(sysconfig.get_config_var("CONFIG_ARGS")):
...     if "x86" in arg.lower():
...         print(arg)
...
--build=x86_64-pc-linux-gnu
build_alias=x86_64-pc-linux-gnu
```

This argument `--build=x86_64-pc-linux-gnu` comes from [this line](https://github.com/termux/termux-packages/blob/master/packages/python/build.sh#L25) of the build script, and the `TERMUX_BUILD_TUPLE` is defined at [this line](https://github.com/termux/termux-packages/blob/0c8a398d0e9eef42c7e533912a0c499fd53d306a/scripts/setup-offline-bundle.sh#L81).

### What steps will reproduce the bug?

Simply run `pip install --user matplotlib==3.5.0` on an aarch64 device.

### What is the expected behavior?

_No response_

### System information

termux-info:

```
Application version:
0.117
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirror.nju.edu.cn/termux/apt/termux-main stable main
# game-repo (sources.list.d/game.list)
deb https://mirror.nju.edu.cn/termux/apt/termux-games games stable
# science-repo (sources.list.d/science.list)
deb https://mirror.nju.edu.cn/termux/apt/termux-science science stable
# unstable-repo (sources.list.d/unstable.list)
deb https://mirror.nju.edu.cn/termux/apt/termux-unstable unstable main
Updatable packages:
dpkg/stable 1.21.1 aarch64 [upgradable from: 1.20.9-5]
gpgv/stable 2.3.3-1 aarch64 [upgradable from: 2.3.3]
termux-keyring/stable 2.4 all [upgradable from: 2.3]
Android version:
10
Kernel build information:
Linux localhost 4.14.117-perf-g2154c22 #1 SMP PREEMPT Thu Jun 4 01:41:59 CST 2020 aarch64 Android
Device manufacturer:
Xiaomi
Device model:
Redmi Note 7 Pro
```

Got it. This is our issue. From the log:

```
  checking for ld used by aarch64-linux-android-clang... no
  configure: error: no acceptable ld found in $PATH
  make: *** [builds/unix/detect.mk:85: setup] Error 1
```

which is due to:

```
$ aarch64-linux-android-clang -print-prog-name=ld
ld

$ ld
bash: ld: command not found
```

So a workaround would be to:

```
$ ln -s lld $PREFIX/bin/ld
```

--- a/packages/matplotlib/Pillow-setup.py.patch
+++ b/packages/matplotlib/Pillow-setup.py.patch
@@ -1,0 +1,28 @@
+--- a/Pillow/setup.py
++++ b/Pillow/setup.py
+@@ -561,8 +561,6 @@
+             or sys.platform.startswith("gnu")
+             or sys.platform.startswith("freebsd")
+         ):
+-            for dirname in _find_library_dirs_ldconfig():
+-                _add_directory(library_dirs, dirname)
+             if sys.platform.startswith("linux") and os.environ.get(
+                 "ANDROID_ROOT", None
+             ):
+@@ -588,16 +586,6 @@
+ 
+         # FIXME: check /opt/stuff directories here?
+ 
+-        # standard locations
+-        if not self.disable_platform_guessing:
+-            _add_directory(library_dirs, "/usr/local/lib")
+-            _add_directory(include_dirs, "/usr/local/include")
+-
+-            _add_directory(library_dirs, "/usr/lib")
+-            _add_directory(include_dirs, "/usr/include")
+-            # alpine, at least
+-            _add_directory(library_dirs, "/lib")
+-
+         if sys.platform == "win32":
+             # on Windows, look for the OpenJPEG libraries in the location that
+             # the official installer puts them

--- a/packages/matplotlib/build.sh
+++ b/packages/matplotlib/build.sh
@@ -1,0 +1,103 @@
+TERMUX_PKG_HOMEPAGE=https://matplotlib.org/
+TERMUX_PKG_DESCRIPTION="A comprehensive library for creating static, animated, and interactive visualizations in Python"
+TERMUX_PKG_LICENSE="custom"
+TERMUX_PKG_LICENSE_FILE="\
+LICENSE/LICENSE
+LICENSE/LICENSE_AMSFONTS
+LICENSE/LICENSE_BAKOMA
+LICENSE/LICENSE_CARLOGO
+LICENSE/LICENSE_COLORBREWER
+LICENSE/LICENSE_JSXTOOLS_RESIZE_OBSERVER
+LICENSE/LICENSE_QHULL
+LICENSE/LICENSE_QT4_EDITOR
+LICENSE/LICENSE_SOLARIZED
+LICENSE/LICENSE_STIX
+LICENSE/LICENSE_YORICK"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=(3.5.1)
+TERMUX_PKG_VERSION+=(1.22.3) # NumPy version
+TERMUX_PKG_VERSION+=(9.0.1)  # Pillow version
+TERMUX_PKG_SRCURL=(https://github.com/matplotlib/matplotlib/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+                   https://github.com/numpy/numpy/archive/refs/tags/v${TERMUX_PKG_VERSION[1]}.tar.gz
+                   https://github.com/python-pillow/Pillow/archive/refs/tags/${TERMUX_PKG_VERSION[2]}.tar.gz)
+TERMUX_PKG_SHA256=(9683da9a0c84d1c42d1bf92ecf6e012d302406a38fd987e3dfbcb7b58b2eea2d
+                   c8f3ec591e3f17b939220f2b9eabb4c5e2db330f8af62c0a3aeee8a4d1a6c0db
+                   01305f0befb644ce7fe90aa0c87573b9163a21d0e65149e8166c24974d9d37d2)
+TERMUX_PKG_DEPENDS="freetype, libc++, libjpeg-turbo, libtiff, libwebp, libxcb, littlecms, openjpeg, python, zlib"
+_PKG_PYTHON_DEPENDS="'cycler>=0.10' 'fonttools>=4.22.0' 'kiwisolver>=1.0.1' 'numpy>=1.17' 'packaging>=20.0' 'pillow>=6.2.0' 'pyparsing>=2.2.1' 'python-dateutil>=2.7'"
+TERMUX_PKG_BUILD_IN_SRC=true
+
+_PYTHON_VERSION=$(. $TERMUX_SCRIPTDIR/packages/python/build.sh; echo $_MAJOR_VERSION)
+
+TERMUX_PKG_RM_AFTER_INSTALL="
+bin/
+lib/python${_PYTHON_VERSION}/site-packages/__pycache__
+lib/python${_PYTHON_VERSION}/site-packages/easy-install.pth
+lib/python${_PYTHON_VERSION}/site-packages/site.py
+"
+
+termux_step_post_get_source() {
+	mv numpy-${TERMUX_PKG_VERSION[1]} numpy
+	mv Pillow-${TERMUX_PKG_VERSION[2]} Pillow
+}
+
+termux_step_pre_configure() {
+	termux_setup_python_crossenv
+	pushd $TERMUX_PYTHON_CROSSENV_SRCDIR
+	_CROSSENV_PREFIX=$TERMUX_PKG_BUILDDIR/python-crossenv-prefix
+	python${_PYTHON_VERSION} -m crossenv \
+		$TERMUX_PREFIX/bin/python${_PYTHON_VERSION} \
+		${_CROSSENV_PREFIX}
+	popd
+	. ${_CROSSENV_PREFIX}/bin/activate
+
+	pushd ${_CROSSENV_PREFIX}/build/lib/python${_PYTHON_VERSION}/site-packages
+	patch --silent -p1 < $TERMUX_PKG_BUILDER_DIR/setuptools-44.1.1-no-bdist_wininst.diff || :
+	popd
+
+	build-pip install numpy
+
+	LDFLAGS+=" -lpython${_PYTHON_VERSION}"
+
+	export NPY_DISABLE_SVML=1
+	pushd $TERMUX_PKG_SRCDIR/numpy
+	pip install .
+	popd
+
+	pushd $TERMUX_PKG_SRCDIR/Pillow
+	INCLUDE=$TERMUX_PREFIX/include LIB=$TERMUX_PREFIX/lib \
+		python setup.py install --force
+	popd
+
+	python setup.py install --force
+}
+
+termux_step_make_install() {
+	export PYTHONPATH=$TERMUX_PREFIX/lib/python${_PYTHON_VERSION}/site-packages
+	python setup.py install --force --prefix $TERMUX_PREFIX
+
+	pushd $PYTHONPATH
+	_MATPLOTLIB_EGGDIR=
+	for f in matplotlib-${TERMUX_PKG_VERSION}-py${_PYTHON_VERSION}-linux-*.egg; do
+		if [ -d "$f" ]; then
+			_MATPLOTLIB_EGGDIR="$f"
+			break
+		fi
+	done
+	test -n "${_MATPLOTLIB_EGGDIR}"
+	popd
+}
+
+termux_step_create_debscripts() {
+	cat <<- EOF > ./postinst
+	#!$TERMUX_PREFIX/bin/sh
+	echo "Installing dependencies through pip. This may take a while..."
+	pip3 install ${_PKG_PYTHON_DEPENDS}
+	echo "./${_MATPLOTLIB_EGGDIR}" >> $TERMUX_PREFIX/lib/python${_PYTHON_VERSION}/site-packages/easy-install.pth
+	EOF
+
+	cat <<- EOF > ./prerm
+	#!$TERMUX_PREFIX/bin/sh
+	sed -i "/\.\/${_MATPLOTLIB_EGGDIR//./\\.}/d" $TERMUX_PREFIX/lib/python${_PYTHON_VERSION}/site-packages/easy-install.pth
+	EOF
+}

--- a/packages/matplotlib/build.sh
+++ b/packages/matplotlib/build.sh
@@ -24,7 +24,7 @@ TERMUX_PKG_SHA256=(9683da9a0c84d1c42d1bf92ecf6e012d302406a38fd987e3dfbcb7b58b2ee
                    c8f3ec591e3f17b939220f2b9eabb4c5e2db330f8af62c0a3aeee8a4d1a6c0db
                    01305f0befb644ce7fe90aa0c87573b9163a21d0e65149e8166c24974d9d37d2)
 TERMUX_PKG_DEPENDS="freetype, libc++, libjpeg-turbo, libtiff, libwebp, libxcb, littlecms, openjpeg, python, zlib"
-_PKG_PYTHON_DEPENDS="'cycler>=0.10' 'fonttools>=4.22.0' 'kiwisolver>=1.0.1' 'numpy>=1.17' 'packaging>=20.0' 'pillow>=6.2.0' 'pyparsing>=2.2.1' 'python-dateutil>=2.7'"
+_PKG_PYTHON_DEPENDS="'cycler>=0.10' 'fonttools>=4.22.0' 'kiwisolver<1.4.0,>=1.0.1' 'numpy>=1.17' 'packaging>=20.0' 'pillow>=6.2.0' 'pyparsing>=2.2.1' 'python-dateutil>=2.7'"
 TERMUX_PKG_BUILD_IN_SRC=true
 
 _PYTHON_VERSION=$(. $TERMUX_SCRIPTDIR/packages/python/build.sh; echo $_MAJOR_VERSION)

--- a/packages/matplotlib/mplsetup.cfg.patch
+++ b/packages/matplotlib/mplsetup.cfg.patch
@@ -1,0 +1,5 @@
+--- a/mplsetup.cfg
++++ b/mplsetup.cfg
+@@ -0,0 +1,2 @@
++[libs]
++system_freetype = true

--- a/packages/matplotlib/setup.py.patch
+++ b/packages/matplotlib/setup.py.patch
@@ -1,6 +1,6 @@
 --- a/setup.py
 +++ b/setup.py
-@@ -316,7 +316,7 @@
+@@ -316,13 +316,13 @@
      setup_requires=[
          "certifi>=2020.06.20",
          "numpy>=1.17",
@@ -9,3 +9,10 @@
          "setuptools_scm_git_archive",
      ],
      install_requires=[
+         "cycler>=0.10",
+         "fonttools>=4.22.0",
+-        "kiwisolver>=1.0.1",
++        "kiwisolver<1.4.0,>=1.0.1",
+         "numpy>=1.17",
+         "packaging>=20.0",
+         "pillow>=6.2.0",

--- a/packages/matplotlib/setup.py.patch
+++ b/packages/matplotlib/setup.py.patch
@@ -1,0 +1,11 @@
+--- a/setup.py
++++ b/setup.py
+@@ -316,7 +316,7 @@
+     setup_requires=[
+         "certifi>=2020.06.20",
+         "numpy>=1.17",
+-        "setuptools_scm>=4",
++        "setuptools_scm<5,>=4",
+         "setuptools_scm_git_archive",
+     ],
+     install_requires=[

--- a/packages/matplotlib/setuptools-44.1.1-no-bdist_wininst.diff
+++ b/packages/matplotlib/setuptools-44.1.1-no-bdist_wininst.diff
@@ -1,0 +1,35 @@
+Copyright (C) 2016 Jason R Coombs <jaraco@jaraco.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the "Software"), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+of the Software, and to permit persons to whom the Software is furnished to do
+so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+--- a/setuptools/command/install_scripts.py
++++ b/setuptools/command/install_scripts.py
+@@ -32,12 +32,7 @@
+         )
+         bs_cmd = self.get_finalized_command('build_scripts')
+         exec_param = getattr(bs_cmd, 'executable', None)
+-        bw_cmd = self.get_finalized_command("bdist_wininst")
+-        is_wininst = getattr(bw_cmd, '_is_running', False)
+         writer = ei.ScriptWriter
+-        if is_wininst:
+-            exec_param = "python.exe"
+-            writer = ei.WindowsScriptWriter
+         if exec_param == sys.executable:
+             # In case the path to the Python executable contains a space, wrap
+             # it so it's not split up.

